components/raftstore-v2/src/operation/command/control.rs

```diff
@@ -0,0 +1,427 @@
+            cmd_type,
+            committed: false,
+            epoch_state,
+            index,
+            delayed_chs: Vec::new(),
+        }
+    }
+
+    pub fn cmd_type(&self) -> AdminCmdType {
+        self.cmd_type
+    }
+
+    /// Delay responding to channel until the command is applied so client won't
+    /// retry with arbitrary timeout.
+    pub fn delay_channel(&mut self, ch: CmdResChannel) {
+        self.delayed_chs.push(ch);
+    }
+
+    /// Same as `delay_channel`, but accepts a batch.
+    pub fn delay_channels(&mut self, chs: Vec<CmdResChannel>) {
+        if self.delayed_chs.is_empty() {
+            self.delayed_chs = chs;
+        } else {
+            self.delayed_chs.extend(chs);
+        }
+    }
+}
+
+/// `ProposalControl` is a rewrite of `CmdEpochChecker` from v1.
+///
+/// Admin command may change the epoch of a region. If a proposal is proposed
+/// after the admin command is proposed but before the command is applied, the
+/// proposal is probably to fail because of epoch not match. `ProposalControl`
+/// aims to detect the failure early. With `ProposalControl`, users can assume
+/// once a command is proposed, it's likely to succeed in the end.
+///
+/// Compared to `CmdEpochChecker`, `ProposalControl` also traces the whole
+/// lifetime of prepare merge.
+pub struct ProposalControl {
+    // Use `LinkedList` to reduce memory footprint. In most cases, the list
+    // should be empty or 1 element. And access speed is not a concern.
+    proposed_admin_cmd: LinkedList<ProposedAdminCmd>,
+    pending_merge_index: u64,
+    term: u64,
+}
+
+impl ProposalControl {
+    pub fn new(term: u64) -> ProposalControl {
+        ProposalControl {
+            proposed_admin_cmd: LinkedList::new(),
+            pending_merge_index: 0,
+            term,
+        }
+    }
+
+    /// Clears all queued conflict callbacks if term changed.
+    ///
+    /// If term is changed, leader is probably changed. Clear all callbacks to
+    /// notify clients to retry with new leader.
+    #[inline]
+    pub fn maybe_update_term(&mut self, term: u64) {
+        match term.cmp(&self.term) {
+            std::cmp::Ordering::Equal => (),
+            std::cmp::Ordering::Greater => {
+                for cmd in mem::take(&mut self.proposed_admin_cmd) {
+                    for cb in cmd.delayed_chs {
+                        apply::notify_stale_req(term, cb);
+                    }
+                }
+                self.term = term;
+            }
+            std::cmp::Ordering::Less => {
+                panic!("term should not decrease, old {}, new {}", self.term, term)
+            }
+        }
+    }
+
+    /// Check if the proposal is conflict with proposed admin commands in
+    /// current term.
+    ///
+    /// Returns None if passing the epoch check, otherwise returns the last
+    /// conflict conflict proposal meta.
+    pub fn check_conflict(
+        &mut self,
+        cmd_type: Option<AdminCmdType>,
```

Comment:
Could you please add a comment for None type cmd_type?  It's not straightforward without looking into the check_conflict implementation